### PR TITLE
Button Component

### LINKED ISSUE #6: Create Button
- [ ] Sizes
- [ ] Colors
- [ ] Icon prop

### PR DESCRIPTION
**Pull request by Deeved**
Was adding to the Button Component:

**Types:**
- primary
- secundary
- ghost
- dashed

**Variation of types:**
- danger

**Sizes**
- sm 
- md
- lg
- xl

**"Behaviors" Styles**
- focus
- disabled
- hover

**Actions**
- click 

**Tests**
- types
- danger
- sizes

fix #6

### CODE REVIEW

File: projects/ion/src/lib/button/button.component.html [hunk 1]
```diff
@@ -1 +1,8 @@
-<button>{{ label || 'configure uma label' }}</button>
+<button
+  [class]="'ion-btn-' + type + ' ion-btn-' + size"
```
Comment (iurynogueira) [inline projects/ion/src/lib/button/button.component.html:2]: @edsonsnunes Check this, you can use this solution to your problem of sizes.

File: projects/ion/src/lib/button/button.component.scss [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+.ion-btn-dashed.danger:hover {
+  background: #FAEAEC;
+  border: 1px dashed #EFA9B0;
+  color: #DE5462;
+}
+
+button:focus {
+  box-shadow: 0px 0px 0px 2px #FCFCFD, 0px 0px 0px 4px #146FF5;
+}
```
Comment (iurynogueira) [inline projects/ion/src/lib/button/button.component.scss:95]: add &:focus in button class.

File: projects/ion/src/lib/button/button.component.scss [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+button:disabled,
+button:disabled:hover,
+button[disabled],
+button[disabled]:hover {
+  background-color: #E4E6EB !important;
+  color           : #AEB2BD !important;
+}
```
Comment (iurynogueira) [inline projects/ion/src/lib/button/button.component.scss:103]: same here

File: projects/ion/src/lib/button/button.component.spec.ts [hunk 4]
```diff
@@ -15,4 +15,85 @@ describe('ButtonComponent', () => {
...
+  });
+
+  it('should render a button with the ion-btn-lg class when size="lg" is passed', async () => {
+    await render(ButtonComponent, {
+      componentProperties: { size: 'lg' },
+    });
+    const button = screen.findByRole('button');
+    const hasClass = (await button).classList.contains(`ion-btn-lg`);
+    expect(hasClass).toBeTruthy();
+  });
```
Comment (iurynogueira) [inline projects/ion/src/lib/button/button.component.spec.ts:80]: Nice! You can improve this separating this in a describe.

File: projects/ion/src/lib/button/button.component.ts [hunk 5]
```diff
@@ -1,10 +1,33 @@
...
+  _disabled = false;
+  _danger = false;
+
+  handleClick() {
+    this.ionOnClick.emit();
+  }
+
+  ngOnInit(): void {
+    this._disabled = this.disabled === 'true' ? true : false;
+    this._danger = this.danger === 'true' ? true : false;
```
Comment (iurynogueira) [inline projects/ion/src/lib/button/button.component.ts:31]: I believe that we can improve this using boolean types to avoid this "cast" 🤔

**Review by iurynogueira — CHANGES_REQUESTED** [review]
Good work! That is great!

**Comment by iurynogueira** [comment]
Don't forget doc in storybook.

**Comment by mizek1** [comment]
You can update your branch and use theme to customize your button. 😄

File: projects/ion/src/lib/button/button.component.scss [hunk 6]
```diff
@@ -1,5 +1,172 @@
...
+  font-family  : 'Source Sans Pro', sans-serif;
+  font-weight  : 600;
+  border-radius: 10px;
+  border       : 0;
+  color        : $neutral-1;
+  cursor: pointer;
+  transition: all 300ms;
+  
+  &:focus {
+    box-shadow: 0px 0px 0px 2px $neutral-1, 0px 0px 0px 4px$primary-5;
```
Comment (iurynogueira) [inline projects/ion/src/lib/button/button.component.scss:17]:
```suggestion
    box-shadow: 0px 0px 0px 2px $neutral-1, 0px 0px 0px 4px $primary-5;
```

File: projects/ion/src/lib/button/button.component.scss [hunk 7]
```diff
@@ -1,5 +1,172 @@
...
+}
+
+.ion-btn-primary.danger {
+  background-color: $negative-6;
+}
+
+.ion-btn-primary.danger:hover {
+  background-color: $negative-5;
+  color: $neutral-1;
+}
```
Comment (iurynogueira) [inline projects/ion/src/lib/button/button.component.scss:85]: use &. Same strategy of button.

File: projects/ion/src/lib/button/button.component.spec.ts [hunk 8]
```diff
@@ -16,3 +16,102 @@ describe('ButtonComponent', () => {
...
+
+describe('Primary ButtonComponent', () => {
+  it('should render a button with the ion-btn-primary class when type="primary" is passed', async () => {
+    await render(ButtonComponent, {
+      componentProperties: { type: 'primary' },
+    });
+    const button = screen.findByRole('button');
+    const hasClass = (await button).classList.contains(`ion-btn-primary`);
+    expect(hasClass).toBeTruthy();
+  });
```
Comment (iurynogueira) [inline projects/ion/src/lib/button/button.component.spec.ts:28]: change to expect to toHaveClass
```suggestion
  it('should render a button with the ion-btn-primary class when type="primary" is passed', async () => {
    await render(ButtonComponent, {
      componentProperties: { type: 'primary' },
    });
    expect(screen.findByRole('button')).ToHaveClass('ion-btn-primary');
  });
```

File: projects/ion/src/lib/button/button.component.spec.ts [hunk 9]
```diff
@@ -16,3 +16,102 @@ describe('ButtonComponent', () => {
...
+describe('Secundary ButtonComponent', () => {
+  it('should render a button with the ion-btn-secundary class when type="secundary" is passed', async () => {
+    await render(ButtonComponent, {
+      componentProperties: { type: 'secundary' },
+    });
+    const button = screen.findByRole('button');
+    const hasClass = (await button).classList.contains(`ion-btn-secundary`);
+    expect(hasClass).toBeTruthy();
+  });
+});
```
Comment (iurynogueira) [inline projects/ion/src/lib/button/button.component.spec.ts:40]: same here.

**Review by iurynogueira — CHANGES_REQUESTED** [review]

File: stories/Button.stories.mdx [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+      label: 'Dashed',
+      type: 'dashed',
+      disabled: true,
+    }}
+  >
+    {Template.bind({})}
+  </Story>
+</Canvas>
+
+- `loading`. Indica que o botão não está preparado para disparar a ação. Como por exemplo, esperando uma resposta do servidor ou preparando e/ou processando algum recurso interno. O evento de clique do botão é desativado quando o load está ativo. É um boolean e opcional. Por padrão é `[loading] = false`.
```
Comment (iurynogueira) [inline stories/Button.stories.mdx:217]:
```suggestion
- `loading`. Indica que o botão não está preparado para disparar a ação. Como por exemplo, esperando uma resposta do servidor ou preparando e/ou processando algum recurso interno. O evento de clique do botão é desativado quando o loading está ativo. É um boolean e opcional. Por padrão é `[loading] = false`.
```

**Review by iurynogueira — CHANGES_REQUESTED** [review]

**Review by mizek1 — CHANGES_REQUESTED** [review]

File: projects/ion/src/lib/button/button.component.ts [hunk 11]
```diff
@@ -1,10 +1,36 @@
...
+  @Input() type?: Type = 'primary';
+  @Input() size?: Size = 'md';
+  @Input() expand? = false;
+  @Input() danger? = false;
+  @Input() disabled? = false;
+  @Input() loading? = false;
+  @Output() ionOnClick? = new EventEmitter();
+
+  handleClick() {
+    if (!this.loading) {
```
Comment (iurynogueira) [inline projects/ion/src/lib/button/button.component.ts:32]: Add a test case to this. "should not call event when is loading"

**Review by iurynogueira — CHANGES_REQUESTED** [review]

File: projects/ion/src/lib/button/button.component.spec.ts [hunk 12]
```diff
@@ -15,4 +27,80 @@ describe('ButtonComponent', () => {
...
+      label: defaultName,
+      ionOnClick: {
+        emit: clickEvent,
+      } as any,
+    });
+    fireEvent.click(button);
+    expect(clickEvent).toHaveBeenCalled();
+  });
+
+  it('should not call event when is loading', async () => {
```
Comment (iurynogueira) [inline projects/ion/src/lib/button/button.component.spec.ts:43]: Nice! 😄

**Review by iurynogueira — APPROVED** [review]

File: projects/ion/src/lib/button/button.component.spec.ts [hunk 13]
```diff
@@ -1,5 +1,17 @@
-import { render, screen } from '@testing-library/angular';
-import { ButtonComponent } from './button.component';
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import { fireEvent, render, screen } from '@testing-library/angular';
+import { ButtonComponent, IonButtonProps } from './button.component';
+
+const defaultName = 'button';
+
+const sut = async (
+  customProps: IonButtonProps = { label: 'button' }
```
Comment (MarcosOliveir4) [inline projects/ion/src/lib/button/button.component.spec.ts:8]:
```suggestion
  customProps: IonButtonProps = { label: defaultName }
```

**Review by edsonsnunes — APPROVED** [review]

**Review by MarcosOliveir4 — CHANGES_REQUESTED** [review]

**Pull request merged**